src/main/java/org/skriptlang/skript/lang/comparator/Comparators.java

```diff
@@ -0,0 +1,263 @@
+	 * This is useful for skipping complex lookups that may require conversion and inversion.
+	 */
+	private static final Map<Integer, Comparator<?, ?>> QUICK_ACCESS_COMPARATORS = Collections.synchronizedMap(new HashMap<>());
+
+	/**
+	 * Registers a new Comparator with Skript's collection of Comparators.
+	 * @param firstType The first type for comparison.
+	 * @param secondType The second type for comparison.
+	 * @param comparator A Comparator for comparing objects of 'firstType' and 'secondType'.
+	 */
+	public static <Type1, Type2> void registerComparator(
+		Class<Type1> firstType,
+		Class<Type2> secondType,
+		Comparator<Type1, Type2> comparator
+	) {
+		Skript.checkAcceptRegistrations();
+
+		if (firstType == Object.class && secondType == Object.class)
+			throw new IllegalArgumentException("It is not possible to add a comparator between objects");
+
+		for (ComparatorInfo<?, ?> info : COMPARATORS) {
+			if (info.firstType == firstType && info.secondType == secondType) {
+				throw new SkriptAPIException(
+					"A Comparator comparing '" + firstType + "' and '" + secondType + " already exists!"
+				);
+			}
+		}
+
+		COMPARATORS.add(new ComparatorInfo<>(firstType, secondType, comparator));
+	}
+
+	/**
+	 * Compares two objects to see if a Relation exists between them.
+	 * @param first The first object for comparison.
+	 * @param second The second object for comparison.
+	 * @return The Relation between the two provided objects.
+	 */
+	@SuppressWarnings("unchecked")
+	public static <Type1, Type2> Relation compare(@Nullable Type1 first, @Nullable Type2 second) {
+		if (first == null || second == null)
+			return Relation.NOT_EQUAL;
+
+		Comparator<Type1, Type2> comparator = getComparator((Class<Type1>) first.getClass(), (Class<Type2>) second.getClass());
+		if (comparator == null)
+			return Relation.NOT_EQUAL;
+
+		return comparator.compare(first, second);
+	}
+
+	/**
+	 * A method for obtaining a comparator that can compare two objects of 'firstType' and 'secondType'.
+	 * Please note that comparators may convert objects if necessary for comparisons.
+	 * @param firstType The first type for comparison.
+	 * @param secondType The second type for comparison.
+	 * @return A comparator capable of determine the {@link Relation} between two objects of 'firstType' and 'secondType'.
+	 * Will be null if no comparator capable of comparing two objects of 'firstType' and 'secondType' was found.
+	 */
+	@Nullable
+	@SuppressWarnings("unchecked")
+	public static <Type1, Type2> Comparator<Type1, Type2> getComparator(Class<Type1> firstType, Class<Type2> secondType) {
+		if (Skript.isAcceptRegistrations())
+			throw new SkriptAPIException("Comparators cannot be retrieved until Skript has finished registrations.");
+
+		int hash = Objects.hash(firstType, secondType);
+
+		Comparator<Type1, Type2> comparator = (Comparator<Type1, Type2>) QUICK_ACCESS_COMPARATORS.get(hash);
+
+		if (comparator == null) {
```

Comment:
There are two scenarios where HashMap#get returns null:
- The map doesn't contain the key, in this case we should compute the result and update the cache
- The map does contain the key, but associated with a null value, in this case we shouldn't call getComparator_i, instead return null immediately (getComparator_i may return null if no comparator found, we should store this in the cache as well).

I think the easiest approach for this is to use computeIfAbsent